### PR TITLE
Move from mocha to karma for running unit tests; add code coverage report generation for npm test

### PR DESCRIPTION
**Pull request by clbond**
* Add code coverage report
* Add some missing typings for react-redux-router, redux-thunk, etc
* Add karma dependencies, remove mocha dependencies

### CODE REVIEW

**Comment by clbond** [comment]
Connected to #112

**Comment by clbond** [comment]
@SethDavenport

File: circle.yml [hunk 1]
```diff
@@ -15,6 +15,7 @@ dependencies:
 
 test:
   post:
+    - cp -R coverage $CIRCLE_ARTIFACTS
```
Comment (SethDavenport) [inline circle.yml]: :+1:

File: karma.conf.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+const postcssInit = require('./webpack/postcss');
+const plugins = require('./webpack/plugins');
+const webpack = require('./webpack.config');
+
+module.exports = (config) => {
+  const coverage = config.singleRun ? ['coverage'] : [];
+
+  config.set({
+    frameworks: [
+      'jasmine',
```
Comment (SethDavenport) [inline karma.conf.js]: what is the rationale for karma + jasmine instead of karma + mocha? Just consistency?

Not opposed, just curious

File: package.json [hunk 3]
```diff
@@ -76,15 +77,26 @@
     "enzyme": "^2.2.0",
     "es5-shim": "^4.5.6",
     "es6-promise": "^3.1.2",
-    "es6-shim": "^0.35.0",
```
Comment (SethDavenport) [inline package.json]: Does babel-polyfill provide everything we need for IE9 support?

e.g. object.assign, array.map/filter/reduce, etc?

The easiest thing would be to make sure the app runs on IE9

File: karma.conf.js (hunk 2 above)
Comment (clbond) [inline karma.conf.js]: To be honest I didn't really give it much thought. Our other projects are using Jasmine so I figured I might as well go with that in order to be consistent. But I don't have a strong opinion one way or the other

**Comment by clbond** [comment]
Whoever merges this, please squash the commits together. Some of the earlier commits had nothing to do with this changeset and were later reversed. For clarity's sake it would be better to squash them all together.

File: package.json [hunk 4]
```diff
@@ -98,13 +110,13 @@
...
+    "react-router-redux": "^4.0.5",
     "redux": "^3.4.0",
     "redux-form": "^5.2.1",
     "redux-localstorage": "^0.4.0",
     "redux-logger": "^2.6.0",
-    "redux-thunk": "^2.0.1",
+    "redux-thunk": "^2.1.0",
     "require-hacker": "^2.1.3",
     "rimraf": "^2.4.3",
     "sinon": "^1.17.3",
```
Comment (SethDavenport) [inline package.json]: If we've switched to Jasmine we can get rid of sinon - Jasmine has its own spies.

Ditto chai - Jasmine includes its own assert/expect syntax.

File: src/components/counter/index.tsx [hunk 5]
```diff
@@ -14,19 +15,18 @@ export default function Counter({
...
+      <Button
+          id="qa-decrement-button"
+          className="bg-black col-2"
+          onClick={decrement}>
         -
       </Button>
 
       <div id="qa-counter-div" className="flex-auto center h1">
-        { counter }
+        {counter.toString()}
```
Comment (SethDavenport) [inline src/components/counter/index.tsx]: why is this needed?

Comment (clbond) [inline src/components/counter/index.tsx]: It's not ... good catch. Removed

File: src/store/configure-store.ts [hunk 6]
```diff
@@ -1,16 +1,22 @@
...
+  createStore,
+  applyMiddleware,
+  compose,
+} from 'redux';
 import { fromJS } from 'immutable';
-const thunk = require('redux-thunk').default;
+import { browserHistory } from 'react-router';
+import { routerMiddleware } from 'react-router-redux';
+
+import thunk from 'redux-thunk';
```
Comment (SethDavenport) [inline src/store/configure-store.ts]: Have the redux-thunk typings been updated? Last time I checked the definity-typed typings were broken for this lib. Specifiy they were incompatible with the official Redux typings.

Granted that was a while ago - hopefully they've updated them.

File: typings.json [hunk 7]
```diff
@@ -9,5 +9,10 @@
     "react-dom": "registry:dt/react-dom#0.14.0+20160412154040",
     "sinon": "registry:dt/sinon#1.16.0+20160517064723",
     "whatwg-fetch": "registry:dt/whatwg-fetch#0.0.0+20160524142046"
+  },
+  "dependencies": {
+    "react-router": "registry:npm/react-router#2.4.0+20160628165748",
```
Comment (SethDavenport) [inline typings.json]: sweet! official typings

File: src/store/configure-store.ts (hunk 6 above)
Comment (SethDavenport) [inline src/store/configure-store.ts]: Just saw the typing.json changes below - ignor the comment above... redux ecosystem typings continie to improve which is awesome!

File: karma.conf.js (hunk 2 above)
Comment (SethDavenport) [inline karma.conf.js]: I'm cool with it

**Comment by SethDavenport** [comment]
Overall looks good - if we can get rid of sinon/chai, and verify that it works on IE9 (there will be some CSS issues with the header but that's ok, there's another issue tracking that) then ok to :shipit:

**Comment by clbond** [comment]
@SethDavenport Alright, I removed the dependencies on chai and sinon and converted all the assertions to Jasmine syntax. OK to merge once CI finishes?

**Comment by clbond** [comment]
@SethDavenport Verified working on IE9. Merging.
